### ISSUE TITLE
Ability for globally averaged output

### ISSUE THREAD

**Issue by bogensch** [E3SM-Project/scream#2738]
It would be nice to have the capability to produce domain/globally averaged output.  My real motivation for this is for DP simulations.  This would allow for high frequency output of profiles while saving substantially on storage requirements and post-processing times.  Currently this is difficult to do in high resolution (or large domain and/or long duration) DP runs because you need to frequently output 3D snapshots and the storage demand can quickly mount.  EXAMPLE: Preparing to submit DP simulations to RCEMIP2 each simulation (of which there are five) produces ~10 TB of data.  More than 95% of this is because of the need to output 3D snapshots to make the domain averaged profiles needed for submission.  Had we this capability, each simulation would produce less than 0.5 TB of data.

Most CRMs (i.e. SAM) have this capability and would be a large quality of life boost for DP users.  I imagine this could also be useful for global simulations as well, but probably less so.

**Comment by PeterCaldwell** [comment]
Can't we do this now by writing the output on a grid which just has a single location and uses mapping weights = 1/cell area? Maybe it is a bit of a hack, but it is something.

**Comment by AaronDonahue** [comment]
I was going to say the same thing.

**Comment by bogensch** [comment]
@PeterCaldwell thanks; once I am able to get a DPv1 case up and running I will try this out and report back here whether or not it was successful.

**Comment by bartgol** [comment]
Yeah, this could be a good IO task. Is the avg supposed to be only in the horiz direction?

**Comment by bogensch** [comment]
Yes, my desire would be for horizontal average only.

**Comment by bartgol** [comment]
Yeah, this would be of course something we can use in combination with other options (avg type, v/h remap, diags...). That said, there is one scenario that _prevents_ using this in a meaningful way: masking. If output is masked, then it makes no sense to compute a global avg. I suppose we could check the input fields, to see if they have a mask field associated to them, or if they have a fillValue associated to them, or... I don't recall all the ways we can get a masked/filled value in a field...

**Comment by bogensch** [comment]
Now that I have prototype DPxx to play with; I went back to attempt to employ the methods suggested by @PeterCaldwell and @AaronDonahue (near the top of the issue).  I now understand that to do this you would need a remapping file.  This is not desirable for DP simulations since domain size / resolution are always different based on user preference and the case selected; thus not desirable to rely on the user to generate this for each instance.  

Thus, for a more elegant and flexible solution, I still advocate my wishlist for the functionality to allow for the option of globally (horizontally) averaged output.  Again, why my motivation is for DP simulations, this could also be beneficial for the global model as well.

(changed the wording of this issue from "domain/globally averaged output" to simply "globally averaged output" as I understand some were confused that I wanted averages of subdomains, which is not the case... and a more complicated issue).

**Comment by bartgol** [comment]
@bogensch Are you ok having _all_ fields in an nc file being globally averaged, or would you expect to have a mix of averaged and spatially varying fields?

Also, I mentioned above masked fields, for which global horiz average makes no sense. I am assuming DP won't generate masked fields, since it has a planar geometry, right? If that's the case, we can simpy error out if we see one (just in case), but assume otherwise that all values are fine.

**Comment by bogensch** [comment]
@bartgol yes absolutely fine with all fields being globally averaged in a particular nc file (that was what I was expecting the final product to be anyway).  

And yes, DP won't generate masked fields.

**Comment by bogensch** [comment]
My need for horizontally averaged output is now motivated by the fact that DPxx is working and is turning out to be an efficient platform for doing very large domain and high resolution simulations.  I am attempting to do what @PeterCaldwell  suggested in this thread (writing the output on a grid which just has a single location and uses mapping weights = 1/cell area).  

Jishi (@jsbamboo) made a proof-of-concept mapping file that works offline on DPxx output (header file attached) for a particular domain layout.  However, when I attempt to do remapping during runtime with this file for the exact same domain layout (my yaml file attached) the run seg faults and doesn't give much useful information for a boy like me (see e3sm.log* file).  

Any idea what could be going on here?  This is our first attempt to use online remapping in a DPxx run.

[dp_mapping_file_header.txt](https://github.com/user-attachments/files/17019937/dp_mapping_file_header.txt)
[e3sm.log.30477310.240915-115118.txt](https://github.com/user-attachments/files/17019939/e3sm.log.30477310.240915-115118.txt)
[scream_hourly_avg_output_pg2_horizavg.yaml.txt](https://github.com/user-attachments/files/17019940/scream_hourly_avg_output_pg2_horizavg.yaml.txt)

**Comment by bartgol** [comment]
@bogensch If you need this feature, we should come up with a special handling of it. Using a map file is somewhat of a temporary patch, since you may end up wasting quite a few threads on GPU. Besides, it requires a new mapping file every time you change the DP resolution, which is nonsense.

@tcclevenger @AaronDonahue @mahf708 (and other I/O devs enthusiasts): this feature should not be hard to add. I wonder if we should make it into a new "remapper", so that we can piggy back on the existing infrastructure (where if a horiz remapper is present, it is run before calling scorpio). Something like "Global2dAverageRemapper" or something along that line. This would allow us to unit test the 2d reduction separately from I/O.

**Comment by ambrad** [comment]
Leaving aside a better internal solution, the map file header doesn't make sense to me, but I might be missing something. Here's what confuses me:
* `nx15` is in the name, but `n_a = 100`. I expected 15^2 = 225.
* `n_s = 3`, which is the number of nonzeros in the mapping matrix. Here again I expected 225. This would describe a matrix A having 1 row, 225 columns, and each entry filled with area/225 or something like that.

Is it possible the map file gives sensible-seeming output offline but in fact is not actually performing the average?

If for some reason we wanted to keep this as a map file, generating a map file would be a one-liner (and fast to run) in the scmlib config script. If you didn't want to rely on a map tool being available, a ~20-line python file could create the necessary nc file.

**Comment by jsbamboo** [comment]
@ambrad nelem = 25 in this case, and ncol = 100 for pg2 output, so n_a = 100. 

S is the sparse matrix coef, not the full n_a x n_b matrix, so it can be much smaller than n_a x n_b. I also think it's strange that n_s is only 3, which is too much smaller than n_a. However, I tested another offline mapping to 15x15 grids (compared to the original resolution 3333m) and then do the fldmean by cdo to get the domain average. The final domain averages from these two mapping files are very close to each other (but not BFB the same as expected).

BTW, I just tested another run using the mapping to 15x15 grids, which gave the same failure using intel cpu. 

The offline remapping from DPxx output to whatever grid is also non-trival as we dont have any source grid description file, and the eamxx/eam outputs dont have lat/lon boundary info, so ncremap can do nothing to infer the source grid SCRIP file (only NCL-ESMF can do that in my experience). If something wrong with the offline mapping file, the referred source SCRIP file could be the problem.

**Comment by jsbamboo** [comment]
@bogensch @ambrad check for the offline remapping:

1. the remapping from dpxx pg2 to 3333m fake lat/lon (NCL-ESMF -> src/dst SCRIP and map, shown in the upper middle) gives similar spatial pattern as the original dpxx pg2 (shown in the left)
2. _although_ the remapping from dpxx pg2 to 1x1 (NCL-ESMF -> src SCRIP, ncremap -> dst SCRIP, ESMF -> map, shown in the bottom middle) has notable difference vs. the step1+fldmean (upper right) (0.41 vs. 0.46 cldfrac_liq at it=77, lev=100), it _works_

<img width="1505" alt="image" src="https://github.com/user-attachments/assets/35b6fa6e-b95b-4fcd-a68e-542f5fd3f6c5">

the visualization steps:
[dpxx_remapping_forncvis.txt](https://github.com/user-attachments/files/17021901/dpxx_remapping_forncvis.txt)

note that the sequence of dimension must be ilev/lev, ncol for ncremap to work correctly

**Comment by ambrad** [comment]
@jsbamboo for the case of pg2 -> 1x1, what map algorithm are you using?

The name of the file suggests bilinear interpolation. That would take the value from just the pg2 cell that is on top of the point at the middle of the 1x1 grid. I would expect that to produce 4, but possibly 3, nonzero entries, which might explain the `n_s = 3` value.

If this is the case, then instead try a finite-volume-to-finite-volume map. Then all pg2 cells will contributed to the 1x1 grid's value and we should see `n_s = 100`.

**Comment by jsbamboo** [comment]
@ambrad yes it's bilin. I tried bilin, neareststod, patch, conserve, conserve2nd, however, conserve and conserve2nd aborted ☹️ 
```
free(): invalid next size (fast)
Program received signal SIGABRT: Process abort signal.
```

it's strange that although n_s is only 3 for bilin,  the final result seems very close to the direct fldmean of the ori pg2 output..

**Comment by jsbamboo** [comment]
@bogensch @ambrad  TR fv2fv also failed: `..EXCEPTION (src/GridElements.cpp, Line 1166) Exodus Grid type is not of type float` 

it seems that conservative interpolation is not possible anyway, as there is no boundary information for the native dpxx outputs? is it possible to output/get the grid corners for dpxx? it would be useful to have the full mesh info (which is available for any ne/RRM grid but not for dpxx)

**Comment by ambrad** [comment]
@jsbamboo thanks for the information. IMO, rather than try to get the tools to give us this simple map, we should do one of two things:
1. make a python tool to generate this simple map;
2. implement a remapper for it.

Either one is simple to do. @bogensch is the full averaging operation meant to be done on model levels, i.e, with no vertical remapping first?

**Comment by jsbamboo** [comment]
@ambrad yes, i tried NCL -- just calculating weights = area / sum(area) works quite well, but the map file format used for ncremap is still troublesome. it requires all nominal variables (yc_a/b, xc_a/b, frac_a/b, area_a/b), although the mapping procedure only needs S, col, row. 

i feel online domain average / mapping is much simpler in dpxx case, since S is easy to calculate

**Comment by bogensch** [comment]
@jsbamboo thanks for your hard work on this so far.

@ambrad yes the full averaging operation is meant to be done on the model levels.  I agree that this is no longer a wishlist item and will change priority. Thanks.

**Comment by ambrad** [comment]
I've assigned @tcclevenger and me but also labeled this "help wanted". When someone can take this, we'll indicate it here.

The task is to create a new horizontal remapper here: https://github.com/E3SM-Project/scream/tree/master/components/eamxx/src/share/grid/remap. It should implement a weighted averaging operation over the whole grid.

@bogensch is the average done like this?
```
    q(lev)_avg = (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i) dp(i,lev))
    dp(lev)_avg = (sum_i area(i) dp(i,lev)) / (sum_i area(i)).
```
As a check, this then gives the global mass of q as
```
    A sum_lev q(lev)_avg dp(lev)_avg
      = A sum_lev [ (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i) dp(i,lev)) x
                    (sum_i area(i) dp(i,lev)) / (sum_i area(i))]
      = A sum_lev [ (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i))]
      = A sum_lev (sum_i area(i) dp(i,lev) q(i,lev)) / A
      = sum_lev sum_i area(i) dp(i,lev) q(i,lev),
```
where A is the global area. Or, rather, should the average ignore dp and not care about consistency with the global mass?

**Comment by tcclevenger** [comment]
@ambrad I'll put this on my list. Try and take a look through the convo this week.

**Comment by bartgol** [comment]
> note that the sequence of dimension must be ilev/lev, ncol so ncremap can work correctly

I thought there was a flag to use, to enable using (ncol,lev) layouts with ncremap. I forgot what it was, maybe `ncremap -P scream <...>`? @czender definitely knows what it is.

**Comment by czender** [comment]
@bartgol `ncremap` will horizontally regrid raw SCREAM files when give the `-P eamxx` option, e.g.,

`ncremap -P eamxx -m ${DATA}/maps/map_ne1024pg2_to_fv256x512_mono.20201201.nc in.nc out.nc`

Under the hood, this invokes ncpdq to permute the dimensions so the horizontal dimension is last. That can take some time.

`ncremap` will also vertically interpolate SCREAM files. It does this without need to permute any dimensions, so it's relatively fast. 

https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/754286611/Regridding+E3SM+Data+with+ncremap#Advanced-Regridding-II%3A-EAMXX-mode

**Comment by jsbamboo** [comment]
thanks for the info! @bartgol @czender

**Comment by bogensch** [comment]
> I've assigned @tcclevenger and me but also labeled this "help wanted". When someone can take this, we'll indicate it here.
> 
> The task is to create a new horizontal remapper here: https://github.com/E3SM-Project/scream/tree/master/components/eamxx/src/share/grid/remap. It should implement a weighted averaging operation over the whole grid.
> 
> @bogensch is the average done like this?
> 
> ```
>     q(lev)_avg = (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i) dp(i,lev))
>     dp(lev)_avg = (sum_i area(i) dp(i,lev)) / (sum_i area(i)).
> ```
> 
> As a check, this then gives the global mass of q as
> 
> ```
>     A sum_lev q(lev)_avg dp(lev)_avg
>       = A sum_lev [ (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i) dp(i,lev)) x
>                     (sum_i area(i) dp(i,lev)) / (sum_i area(i))]
>       = A sum_lev [ (sum_i area(i) dp(i,lev) q(i,lev)) / (sum_i area(i))]
>       = A sum_lev (sum_i area(i) dp(i,lev) q(i,lev)) / A
>       = sum_lev sum_i area(i) dp(i,lev) q(i,lev),
> ```
> 
> where A is the global area. Or, rather, should the average ignore dp and not care about consistency with the global mass?

@ambrad your formulation looks good to me.  When I compute the horizontal average offline I typically neglect dp, but I feel it should be included in our horizontal remapper.

**Comment by bartgol** [comment]
> > this feature should not be hard to add. I wonder if we should make it into a new "remapper", so that we can piggy back on the existing infrastructure (where if a horiz remapper is present, it is run before calling scorpio). Something like "Global2dAverageRemapper" or something along that line. This would allow us to unit test the 2d reduction separately from I/O.
> 
> Why not a reducer in a generic diagnostic such that MyVarGlobalAvg returns MyVar globally averaged? We already allow 3D (time,ncol,lev) and 2D (time,ncol) variables together in the same stream, we should also allow 1D (time) and 2D (time,lev) as well

Besides the fact that @bogensch said he's interested in having *all* vars in the file being glob integrals, I think a remapper makes things easier, since we can do the MPI stuff in one sweep. Diags are (as of now) computed one by one, with diag2 being computed only after diag1 has been fully computed. Since global avg requires MPI, that means N global averages will require N MPI calls (with N device fences). On the other hand, if you implement a "all-or-nothing" approach, you can do the local reduction for all fields, one barrier, and one MPI reduction.

Perhaps these are minor points, since most of the run time is spent somewhere else. Alas, unless we are really thinking of having global-avg fields in the same file as non-global-avg fields, we may as well go the slightly more performant way..

**Comment by bartgol** [comment]
We can def add a "GlobalHorizAverage" diagnostic, maybe requestable via "my_field_2d_avg" or something. That would make users able to add a handful of 2d averages to an existing stream.

**Comment by bogensch** [comment]
An update that thanks to the efforts of @AaronDonahue and @jsbamboo, we are now able to produce horizontally averaged output in DPxx via the remap method (i.e. making a mapping file appropriate for the geometry of your DP domain and feeding that in your YAML file).  I am currently working on a streamlined user friendly python script that will quickly generate these mapping files for users where they only need to input basic geometry info for their domain (and I will upload prefab mapping files for standard domain geometries to the scmlib repo).  

Thus, now that we have the ability to get horizontally averaged output I am happy.  While I still see benefits for putting this functionality into the code, I no longer see it as a big priority since there is a path I can do this (which actually won't be that burdensome) and it is not blocking me in doing my ambitious simulations anymore.  Thus, I'm probably going to reduce the priority of this issue back to wishlist.

**Comment by bartgol** [comment]
@bogensch Thanks for the update! I realized I had an old branch with an initial impl. I just rebased it on master and pushed it. I (or someone else) may soon add a few unit tests and maybe put up a PR.

**Comment by bartgol** [comment]
> @bartgol I am interested in taking your code to completion with a PR (so feel free to punt or assign me)

if you are interested, the branch is `bartgol/eamxx/havg-diag`. I only implemented the diag. You'd still need to a) add support for it in AtmosphereOutput (including adding it to the factory), and b) testing.

I implemented it as a diag, even though if one wants to avg ALL fields in a stream, we may be better off handling it in the I/O class (in order to do one MPI reduction for all fields). But I/O is already complicated, so I thought I'd go with the easy-to-impl way.

**Comment by bartgol** [comment]
> Sounds good, I think I can deal with it (and it will be a good thing for me to try something with MPI reduce)
> 
> > Edit: I think my diag is incomplete, since I'm not dividing by ncols.
> 
> @bartgol to be strictly correct, we should use the column area as weighting factor; do we happen to have that information in the grid or somewhere handy? Or do we have to calculate it?

Yes, we do. We can compute the global area during set_grids (when we do have access to the grid), and then use it in run_impl.

**Comment by bartgol** [comment]
This has been taken care of in E3SM-Project/E3SM#6776. Closing.
